Exclude URL delimiters from mapped chars

jlledom opened the pull request:
**What this PR does / why we need it**:

Some queries to manticore, for the `proxy_rule` index, are inconsistent and don't return all the matching results. In particular, the Jira issue includes an example with the keyword `foto`.

I tried locally and reproduced the bug, for both sphinx and manticore.

I've been able to solve this by removing some allowed characters from the index. I'm not sure how this affects manticore internals but the truth is having the character `/` as allowed causes the bug.

As I understood it, [charset_table](https://sphinxsearch.com/docs/current/conf-charset-table.html) means the list of characters that are considered valid characters inside a search query. Characters not in `charset_table` will be considered "delimiters" and are ignored in queries.

Since mapping rules are URLs, I think it's correct to accept only URL characters, and only those which are not reserved delimiters according to [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986#section-2.2)

According to [this SO answer](https://stackoverflow.com/questions/7109143/what-characters-are-valid-in-a-url). Those characters are:

```
A-Z, a-z, 0-9, -, ., _, ~, :, /, ?, #, [, ], @, !, $, &, ', (, ), *, +, ,, ;, %, =
```

But from those, some are reserved delimiters, so I allowed these ones in our manticore index:

```
A-Z, a-z, 0-9, -, ., _, ~, %
```

This solves the issue, though I don't entirely understand why it failed before. I mean, `/` was allowed as a character to be considered in searches, so a query like `/fotos` should work if it actually matches the pattern, which was the problem with that? I looks correct to me.

**Which issue(s) this PR fixes** 

https://issues.redhat.com/browse/THREESCALE-10930

**Verification steps** 

1. Add the next mapping rules:
```
/biometria-servico/services/eleitoral/fotos
/biometria-servico/services/fotos/eleitoral
/bar/v1/foto/fotos
/bar/v1/something-else/fotos
/bar/v1/something/else/fotos
/bar/v1/someting/else/this/is/even/larger/than/the/one/failing/fotos
/biometria-servico/services/eleitora/fotos
/biometria-servico/services/eleitor/fotos
/biometria-servico/services/eleitoral/a/fotos
/biometria-servico/services/eleitorala/fotos
/biometria-servico/services/eleitorb/fotos
/biometria-servico/services/eleitorbl/fotos
/biometria-servico/services/ases/codObjetoAse/fotos-miniatura
/biometria-servico/services/ases/{codObjetoAse}/fotos-miniatura
/biometria-servico/services/ases/{codObjetoAse}/miniatura-fotos
/biometria-servico/services/eleitoral/biometria/v1/ase/{codObjetoAse}/miniatura-fotos
```
2. Search for `fotos`
3. On `master`, it will return 10 results, on this branch, it will return all 16 results.

mayorova commented:
For reference, these are the patterns that are currently found in `master`:
| Pattern                                                                                | `master` |
| ----                                                                                   | ----         |
| /biometria-servico/services/eleitoral/fotos                                            |              |
| /biometria-servico/services/fotos/eleitoral                                            |  X           |
| /bar/v1/foto/fotos                                                                     |  X           |
| /bar/v1/something-else/fotos                                                           |  X           |
| /bar/v1/something/else/fotos                                                           |  X           |
| /bar/v1/someting/else/this/is/even/larger/than/the/one/failing/fotos                   |              |
| /biometria-servico/services/eleitora/fotos                                             |  X           |
| /biometria-servico/services/eleitor/fotos                                              |  X           |
| /biometria-servico/services/eleitoral/a/fotos                                          |              |
| /biometria-servico/services/eleitorala/fotos                                           |              |
| /biometria-servico/services/eleitorb/fotos                                             |  X           |
| /biometria-servico/services/eleitorbl/fotos                                            |              |
| /biometria-servico/services/ases/codObjetoAse/fotos-miniatura                          |              |
| /biometria-servico/services/ases/{codObjetoAse}/fotos-miniatura                        |  X           |
| /biometria-servico/services/ases/{codObjetoAse}/miniatura-fotos                        |  X           |
| /biometria-servico/services/eleitoral/biometria/v1/ase/{codObjetoAse}/miniatura-fotos  |  X           |

I mean... why is `/biometria-servico/services/eleitorb/fotos` being found, but not `/biometria-servico/services/eleitorala/fotos` ??? How are they different in terms of tokenization?...

akostadinov commented:
This is very interesting. Awesome table @mayorova . Is it possible that with `/` some words become too long and thus are not fully tokenized?

jlledom commented:
> Nice @jlledom !
> 
> But I also don't quite understand the logic... I mean, OK, if `/` was considered a "valid character", probably, indeed searching for "fotos" would not return the desired results, because "fotos" would not be an indexed token, and could only be searched as part of a longer query. But what I don't understand is - why would it still work for some paths then!??

Yeah, I'm confused too...

> On the other hand, this character set was previously added explicitly, in order to fix some other failing searches, I think: https://github.com/3scale/porta/pull/2138/files

I didn't know about that PR, but apparently the problem it solves is also solved in this branch. I tried searching for:

```
{codObjetoAse}
/biometria-servico/services/eleitoral/biometria/v1/ase/{codObjetoAse}/miniatura-fotos
/fotos-miniatura
/bar/v1/
```

And everything seems to work as expected. The only maybe wrong case I found is searching for:

```
/biometria-servico/services/eleito
```

Which returns the 8 results that are exactly equal but also this additional result:

```
/biometria-servico/services/fotos/eleitoral
```

Which is not exact but I think it's pretty good anyway.

> Also, we need to also take care about such characters as `{}`, because, while they are not standard URLs, they are common in the mapping rules.

Yeah but removing them from index is not important I think. For instance, we have an url including `{codObjetoAse}`. Then, if you search exactly for the same string `{codObjetoAse}` it will return the correct results because the brackets are ignored.

jlledom commented:
> Is it possible that with `/` some words become too long and thus are not fully tokenized?

The problem could be the token size.

This one is not returned by master:

```
/biometria-servico/services/eleitoral/fotos
```

But this one, just one character shorter, is returned:

```
/biometria-servico/services/eleitora/fotos
```

On the other hand, this one, much larger, is returned:

```
/biometria-servico/services/eleitoral/biometria/v1/ase/{codObjetoAse}/miniatura-fotos
```

But that might be because the brackets are not indexed in `master`, so that larger string probably generates several tokens.

However, this portion:
```
/biometria-servico/services/eleitoral/biometria/v1/ase/
```

should be a token itself since all it's characters are indexed, but is larger than some of the results not returned.

Soooo.... `¯\_(ツ)_/¯`

akostadinov commented:
@jlledom , I mean that the pattern appears after the maximum word length. If word appears before, then the overall length not an issue.

jlledom commented:
> @jlledom , I mean that the pattern appears after the maximum word length. If word appears before, then the overall length not an issue.

That makes sense. Maybe that's it?

mayorova commented:
> @jlledom , I mean that the pattern appears after the maximum word length. If word appears before, then the overall length not an issue.

Can you give an example please @akostadinov ?

@jlledom by the way, as we are at it, how about we bump `min_infix_len` ? I think this is the only index that has it set to `1`, and I believe this setting is not even supported.
https://manual.manticoresearch.com/Creating_a_table/NLP_and_tokenization/Wildcard_searching_settings#min_infix_len

> The minimum allowed non-zero value is 2.

The other indexes have it set to 3.

akostadinov commented:
> Can you give an example please @akostadinov ?

I just noticed from your examples that almost the same examples with 1 character or longer difference don't show up. In Joan's message there is an example with two values. I spent half an hour searching but didn't find any configuration option that would limit the word length. I see nothing in [docs](https://manual.manticoresearch.com/Creating_a_table/NLP_and_tokenization/Wildcard_searching_settings).

I think increasing `min_infix_len` makes sense. Maybe the thinking was that in apis we often have `/p/whatever` or `/xz/whatever`. Maybe that's why they set to 1. But still 1 character doesn't make sense especially if we allow `/`.

jlledom commented:
> @jlledom by the way, as we are at it, how about we bump `min_infix_len` ?
> 
> The other indexes have it set to 3.

I increased it to 3: https://github.com/3scale/porta/pull/4021/commits/f45c029463ab24d3e5ad6a50687464c9e81f399b

I also updated the test suite: https://github.com/3scale/porta/pull/4021/commits/b74882b59430bcc2c5a507528f1c22c6249620df

It was checking for some allowed characters that I removed. I also made small changes in what the tests index and search for, because I think the old way was not compatible with `min_infix_len=3` and in general they make more sense now IMO.

akostadinov commented:
@jlledom but still `/` is not used. Will the exact search like in #2138 work now?

jlledom commented:
> @jlledom but still `/` is not used. Will the exact search like in #2138 work now?

I mentioned that in [this comment](https://github.com/3scale/porta/pull/4021#issuecomment-2669289403). I think it'll find the exact match, and maybe some other additional partial matches.


It will use `/` as a separator for tokens, for instance, we index `/path/to/resource`, then it will ignore the slashes and tokenize `path`, `to` and `resource`.

Then you search for `/path/to/resource` and it will do the same, ignore the slashes and look for patterns containing `path`, `to` and `resource`, it will find the exact match `/path/to/resource` because it's the best match, but it might add some other results like `/path/resource/to`. See my example in the comment I linked above.

akostadinov commented:
So my question was whether it in fact does work like this. Can we have a test searching for something like "api/path1/path2/"?

jlledom commented:
> So my question was whether it in fact does work like this. Can we have a test searching for something like "api/path1/path2/"?

Don't quite understand the test you suggest. What would the test index, search for, and expect? Tell me and I'll write it.

akostadinov commented:
Take a few from your example. e.g.
```
/bar/v1/something-else/fotos
/bar/v1/something/else/fotos
/bar/v1/someting/else/this/is/even/larger/than/the/one/failing/fotos
/biometria-servico/services/eleitora/fotos
/biometria-servico/services/eleitor/fotos
```
then search for `/one/failing/fotos`

jlledom commented:
@akostadinov I added the test you request, and in the process, I found that in fact it didn't work as I expected, so I had to make a few changes: https://github.com/3scale/porta/pull/4021/commits/aeb4d51e9a0639f3eb1da7af6ba991a32bf794e6

I remove the `star: true` options because it transformed the query into something like this:

```
SELECT * FROM `proxy_rule_core` WHERE MATCH('\\/*bar*\\/*v1*\\/*foto*\\/*fotos*')
```

So it added the stars to every keyword, and I don't know why, that always returned empty results.

What I did is to enable the [expand_keywords](https://manual.manticoresearch.com/Creating_a_table/NLP_and_tokenization/Wildcard_searching_settings#expand_keywords) option which allows to search without stars and then add the stars on the fly. Now the query looks like this:

```
SELECT * FROM `proxy_rule_core` WHERE MATCH('\\/bar\\/v1\\/foto\\/fotos')
```

Now I wonder if I should enable `index_exact_words`, see this [section](https://manual.manticoresearch.com/Creating_a_table/NLP_and_tokenization/Wildcard_searching_settings#min_prefix_len) in the docs, WDYT?. IMO results seem to be pretty solid already.

On the other hand, I also made some modifications in the test suite: https://github.com/3scale/porta/pull/4021/commits/2b275a4390fd0d63dbbf07cb23c531a5fbfdefe6

I added the suggested test, but I wanted to verify the exact match was being return first. Unfortunately, we were using `ProxyRuleQuery` to get the results, which alters the order returned from Sphinx. Since this test suite is for the index itself, I thought we better bypass `ProxyRuleQuery` query sphinx directly from tests, this way we can verify the order of results given by Sphinx. We already have [another test suite](https://github.com/3scale/porta/blob/master/test/unit/queries/proxy_rule_query_test.rb) for `ProxyRuleQuery` anyway.

In order to not duplicate code, I extract the search options in `ProxyRuleQuery` to a constant which I read from the tests.

jlledom commented:
> Take a few from your example. e.g.
> 
> ```
> /bar/v1/something-else/fotos
> /bar/v1/something/else/fotos
> /bar/v1/someting/else/this/is/even/larger/than/the/one/failing/fotos
> /biometria-servico/services/eleitora/fotos
> /biometria-servico/services/eleitor/fotos
> ```
> 
> then search for `/one/failing/fotos`

Yeah, I wrote a test like that. With the new code, this is what your example returns:

![image](https://github.com/user-attachments/assets/7cda2067-aa0d-4f08-86b7-5d37d12c0bbc)

akostadinov commented:
Yeah, I think exact words matching should be nice. Do we have a test for the simeple `fotos` use case for example? Whether it still matches partial?

While on it, I don't understand what you mean with "add the stars on the fly".

jlledom commented:
> Also can we avoid reordering by activerecord? But maybe not in this PR, just asking.

It's not trivial I think, check the code:

https://github.com/3scale/porta/blob/aeb4d51e9a0639f3eb1da7af6ba991a32bf794e6/app/queries/proxy_rule_query.rb#L15-L22

Notice the call to `order_by`. No matter which scope you set, it will always add the `ORDER BY` clause to the query. Even if `@sort` and `@direction` are `nil`, there are some default values hardcoded in the model:

https://github.com/3scale/porta/blob/f9589a114d77a2f141ee415ee0e0e4c22275d649/app/models/proxy_rule.rb#L23-L25

So it requires a bit of work to remove without breaking anything.

jlledom commented:
@akostadinov I added the test you requested: https://github.com/3scale/porta/pull/4021/commits/e4c3e0acfae5f4279f40efc6ef727c2e54120ba7

> Maybe we can add `reorder(...)` whenever full text search was used, see https://stackoverflow.com/a/396771/520567

I guess you mean `FIELD()`, right? I don't like it because it's probably a MySQL-only function. 

The UI allows to reorder by columns, but if not column has selected, then it orders by position ASC. That's a bit weird because the UI doesn't indicate the results are ordered by position, so it allows you to click on the position column... which doesn't change anything because it was already ordered. IMO, if the user searches for a pattern, it expects to get the best matching results first, so we should order by best match by default, and let the user reorder by columns if they want.

I would do that in another PR anyway.

akostadinov commented:
Thanks. Will check it out! 
just FYI, perhaps this is an ugly but more portable approach https://stackoverflow.com/a/9475755/520567

jlledom commented:
> Thanks. Will check it out! just FYI, perhaps this is an ugly but more portable approach https://stackoverflow.com/a/9475755/520567

Definitely ugly hehe. If we wanted to order by best match, I'm sure Sphinx has some SQL syntax to do that, no need to do these tricky thing I guess.

mayorova commented on app/queries/proxy_rule_query.rb:
```diff
@@ -1,6 +1,10 @@
 
 class ProxyRuleQuery
+  DEFAULT_SEARCH_OPTIONS = {
+    ids_only: true, per_page: ThreeScale::Search::Helpers::SPHINX_PAGE_SIZE_INFINITE, ignore_scopes: true
```
I understand that `star: true` was removed, and replaced with `set_property expand_keywords: 1`. But how is the behavior different between these two options?

jlledom replied:
On manticore internals I got more questions than answers, but this is my guess:

Before, we searched for `*fotos*`, and now we search for `(fotos | *fotos* | =fotos)` ([docs](https://manual.manticoresearch.com/Creating_a_table/NLP_and_tokenization/Wildcard_searching_settings#expand_keywords)), so it will return more results than before because there are three clauses joined by `OR`.

Now I wonder: why is `fotos` not equal to `*fotos*`? The only reason is because `fotos` means "exact match" and `*fotos*` means "partial match"; but then, why is `fotos` not equal to `=fotos`?

Mysteries of life.

jlledom replied:
This to keep compatibility with https://github.com/3scale/porta/pull/2138. There's a difference when we try to search for exact URLs like `/bar/v1/foto/fotos`. When `star` is true and `expand_keywords`  is not set, the query is transformed to:

```
SELECT * FROM `proxy_rule_core` WHERE MATCH('\\/*bar*\\/*v1*\\/*foto*\\/*fotos*')
```

And that returns no results. If you remove the `star` option, it only looks for exact matches:

```
SELECT * FROM `proxy_rule_core` WHERE MATCH('\\/bar\\/v1\\/foto\\/fotos')
```

And that returs only the exact match:

```
+------+--------------------+----------------+-------------------+--------------------+----------+-----------------------+------------+
| id   | pattern            | sphinx_deleted | sphinx_updated_at | sphinx_internal_id | owner_id | sphinx_internal_class | owner_type |
+------+--------------------+----------------+-------------------+--------------------+----------+-----------------------+------------+
|  186 | /bar/v1/foto/fotos |              0 |        1740732171 |                 10 |        2 | ProxyRule             | Proxy      |
+------+--------------------+----------------+-------------------+--------------------+----------+-----------------------+------------+
```

If you also set `expand_keywords = 1` then the same query (without stars) returns also partial matches, but putting the exact match first:

```
+------+----------------------------------------------------------------------+----------------+-------------------+--------------------+----------+-----------------------+------------+
| id   | pattern                                                              | sphinx_deleted | sphinx_updated_at | sphinx_internal_id | owner_id | sphinx_internal_class | owner_type |
+------+----------------------------------------------------------------------+----------------+-------------------+--------------------+----------+-----------------------+------------+
|  186 | /bar/v1/foto/fotos                                                   |              0 |        1740732392 |                 10 |        2 | ProxyRule             | Proxy      |
|  204 | /bar/v1/something-else/fotos                                         |              0 |        1740732392 |                 11 |        2 | ProxyRule             | Proxy      |
|  222 | /bar/v1/something/else/fotos                                         |              0 |        1740732392 |                 12 |        2 | ProxyRule             | Proxy      |
|  240 | /bar/v1/someting/else/this/is/even/larger/than/the/one/failing/fotos |              0 |        1740732392 |                 13 |        2 | ProxyRule             | Proxy      |
+------+----------------------------------------------------------------------+----------------+-------------------+--------------------+----------+-----------------------+------------+
```

And the original search for `fotos` still returns all relevant results:

```
mysql> SELECT * FROM `proxy_rule_core` WHERE MATCH('fotos') AND `owner_type` = 'Proxy' AND `owner_id` = 2 AND `sphinx_deleted` = 0 LIMIT 0, 1000000 OPTION cutoff=0;
+------+---------------------------------------------------------------------------------------+----------------+-------------------+--------------------+----------+-----------------------+------------+
| id   | pattern                                                                               | sphinx_deleted | sphinx_updated_at | sphinx_internal_id | owner_id | sphinx_internal_class | owner_type |
+------+---------------------------------------------------------------------------------------+----------------+-------------------+--------------------+----------+-----------------------+------------+
|  366 | /biometria-servico/services/ases/codObjetoAse/fotos-miniatura                         |              0 |        1740732392 |                 20 |        2 | ProxyRule             | Proxy      |
|  384 | /biometria-servico/services/ases/{codObjetoAse}/fotos-miniatura                       |              0 |        1740732392 |                 21 |        2 | ProxyRule             | Proxy      |
|  402 | /biometria-servico/services/ases/{codObjetoAse}/miniatura-fotos                       |              0 |        1740732392 |                 22 |        2 | ProxyRule             | Proxy      |
|  420 | /biometria-servico/services/eleitoral/biometria/v1/ase/{codObjetoAse}/miniatura-fotos |              0 |        1740732392 |                 23 |        2 | ProxyRule             | Proxy      |
|  150 | /biometria-servico/services/eleitoral/fotos                                           |              0 |        1740732392 |                  8 |        2 | ProxyRule             | Proxy      |
|  168 | /biometria-servico/services/fotos/eleitoral                                           |              0 |        1740732392 |                  9 |        2 | ProxyRule             | Proxy      |
|  186 | /bar/v1/foto/fotos                                                                    |              0 |        1740732392 |                 10 |        2 | ProxyRule             | Proxy      |
|  204 | /bar/v1/something-else/fotos                                                          |              0 |        1740732392 |                 11 |        2 | ProxyRule             | Proxy      |
|  222 | /bar/v1/something/else/fotos                                                          |              0 |        1740732392 |                 12 |        2 | ProxyRule             | Proxy      |
|  240 | /bar/v1/someting/else/this/is/even/larger/than/the/one/failing/fotos                  |              0 |        1740732392 |                 13 |        2 | ProxyRule             | Proxy      |
|  258 | /biometria-servico/services/eleitora/fotos                                            |              0 |        1740732392 |                 14 |        2 | ProxyRule             | Proxy      |
|  276 | /biometria-servico/services/eleitor/fotos                                             |              0 |        1740732392 |                 15 |        2 | ProxyRule             | Proxy      |
|  294 | /biometria-servico/services/eleitoral/a/fotos                                         |              0 |        1740732392 |                 16 |        2 | ProxyRule             | Proxy      |
|  312 | /biometria-servico/services/eleitorala/fotos                                          |              0 |        1740732392 |                 17 |        2 | ProxyRule             | Proxy      |
|  330 | /biometria-servico/services/eleitorb/fotos                                            |              0 |        1740732392 |                 18 |        2 | ProxyRule             | Proxy      |
|  348 | /biometria-servico/services/eleitorbl/fotos                                           |              0 |        1740732392 |                 19 |        2 | ProxyRule             | Proxy      |
+------+---------------------------------------------------------------------------------------+----------------+-------------------+--------------------+----------+-----------------------+------------+
16 rows in set (0,00 sec)
--- 16 out of 16 results in 0ms ---
```

See my comment above https://github.com/3scale/porta/pull/4021#issuecomment-2671499884

mayorova replied:
@jlledom Great info, thanks for the details.

akostadinov commented on test/unit/indices/proxy_rules_indices_test.rb:
```diff
@@ -7,132 +7,65 @@ class ProxyRulesIndicesTest < ActiveSupport::TestCase
-      end
-    end
+  def search_for(pattern, **options)
+    search_options = ProxyRuleQuery::DEFAULT_SEARCH_OPTIONS.merge options
+    ProxyRule.search(ThinkingSphinx::Query.escape(pattern), search_options).to_a
```
This is great, I'm just wondering, can we use `scoped_search`, what we normally use in our code? e.g.
```suggestion
    ProxyRule.scoped_search(by_query: pattern).to_a
```

idk what these options are so perhaps directly accepting this will break everything.

jlledom replied:
What's the difference between `search` and `scoped_search`? I've been looking but I couldn't find any occurrence of `scoped_search` in our code or in `thinking-sphinx` code

akostadinov replied:
sorry, it is `scope_search` , used everywhere I believe. I don't think we use plain `search`.

jlledom replied:
But what's the difference between them? The code was already calling `search`, I didn't change that.

`search` is a `ThinkingSphinx` method, but `scope_search` is ours. I read the code and couldn't understand it.

akostadinov replied:
The point is that when useing the method directly, you also directly escape the string. So if we change escaping in `serach_scope` that will not automatically reflect in the test. Anyway, not a big deal.

jlledom commented:
> The problem is not manticore. It would order the result. But then in SQL we use the found IDs like `WHERE id IN (...result from manticore...)`. But the SQL server doesn't care about the order of that list. It only cares whether it's rows match the condition or not.

Yeah you're right, I don't know how to solve this. Would it be possible to delegate pagination to sphinx rather than AR? That way at least most relevant results would appear in the first page.

mayorova commented:
while we are changing the sphinx config, how about getting rid of the following warning:


```
WARNING: key 'workers' is deprecated in /etc/manticoresearch/manticore.conf line 12; use 'default value' instead.
```

This worked for me:
```
$ git diff config/thinking_sphinx.yml
diff --git a/config/thinking_sphinx.yml b/config/thinking_sphinx.yml
index ae8de87d6..43318b893 100644
--- a/config/thinking_sphinx.yml
+++ b/config/thinking_sphinx.yml
@@ -58,6 +58,9 @@ base: &default
   log: <%= ENV['THINKING_SPHINX_LOG'] %>
   <% end %>
 
+  # This searchd config is deprecated in manticore, see https://github.com/pat/thinking-sphinx/issues/1254
+  workers: null
+
 development:
   <<: *default
```

Reference: https://github.com/pat/thinking-sphinx/issues/1254

akostadinov commented:
@mayorova , yes, we can remove `workers` option. 

>  I don't know how to solve this

One options is to use ugly SQL.

> Would it be possible to delegate pagination to sphinx rather than AR?

Surely manticore supports paginating. It will require some good refactoring to do that though I think. Because we're using some gem for pagination. Maybe we'll even have to hook into internals.

mayorova commented:
@jlledom @akostadinov There is an insteresting think I've found: https://edgeapi.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-in_order_of

I think this could be useful if we wanted to order by the list of IDs, as returned by manticore. But, I think it's confusing, because we also support ordering by a selected column etc. 

And I think this is also what complicates delegating pagination to search engine - the user might request ordering by something else.

mayorova commented:
I've been checking locally how we could possibly apply the configuration change on a manticore server, and here is what I found out (only relevant for the specific change in this PR).

1. Good news - when the new configuration is applied, the search keeps working for all of the indexes, except the one affected. And in fact, the proxy rules search only stops working when the configuration file is updated "on the go" - when the search server is running. When the server is restarted (with the new configuration), the search for proxy rules starts working - but in the same way it worked before the change. 

2. There is a possibility to clear and reindex only a specific index via Rake task by providing a specific index name in an env var: 
```
INDEX_FILTER=proxy_rule_core bundle exec rake ts:rt:clear
INDEX_FILTER=proxy_rule_core bundle exec rake ts:rt:index
```

However, there is a caveat... In order for `clear` task to work correctly, the server needs to be stopped, otherwise it seems that it doesn't have an effect.

So, the workflow to apply the changes could be the following:

1. Stop the server
2. Run `ts:rt:clear` task for `proxy_rule_core`
3. Start the server
4. Run `ts:rt:index` task for `proxy_rule_core`

I believe it would be the same workflow also for the rollback scenario.

Now, the problem (for on-premises specifically) is that in OpenShift manticore and system pods are completely disconnected, and the system pod (which is the one that would run the rake task) can not access the searchd's volume. And so, `ts:rt:clear` task would not even work, I believe, because it would not be able to access the required path.

But actually, the rake task is not that necessary, because we know that the only thing it does is removing some files under the manticore's index data folder.
So, I guess we would need to have opearator do something like:
- shut down manticore pod
- spin up a new pod based on the system image **with manticore volume attached** 
- delete the files related to the updated search index (if the new pod actually is based on system image, we can run the rake task)
- start manticore pod
- run `rt:index` in the new pod for reindexing the proxy rules index

Then this temporary pod could be shut down.

Probably this workflow can be reused for any futher upgrades, we could just control which indexes need to be reindexed.

akostadinov commented:
> I think this could be useful if we wanted to order by the list of IDs, as returned by manticore. But, I think it's confusing, because we also support ordering by a selected column etc.
> 
> And I think this is also what complicates delegating pagination to search engine - the user might request ordering by something else.

When user does full text search, the results should always order by relevance. I don't think anything else makes sense.
 
>    Stop the server
>    Run ts:rt:clear task for proxy_rule_core
>    Start the server
>    Run ts:rt:index task for proxy_rule_core

This can only work for a local sphinx/manticore and we don't run like this. We need an SQL command, probably `drop table`. Or did it work for you on a containerized manticore?

> Now, the problem (for on-premises specifically)...

I highly suggest we do full reindexing on-prem, i.e. clear the manticore volume. It will not take that much time there and customers will know what's going on. There upgrades are epected to have a little downtime I think.

mayorova commented:
> When user does full text search, the results should always order by relevance. I don't think anything else makes sense.

Well, in our case I think it does make sense. First, because it's not a real full-text search, in the sense that the indexed "texts" are really a few words phrases at most, and the search is typically performed by a single keyword, so we really just search for a complete match most of the time. And second, the user might actually be interested in showing the results in "created at" order, or "latest traffic".
I mean, I actually never heard of a requirement (for 3scale, I mean) about sorting the results in order of relevance :shrug: (which of course doesn't mean that nobody has ever requested this or thought about this).
What I'm trying to say - let's keep it as it is :grimacing:  

> This can only work for a local sphinx/manticore and we don't run like this. We need an SQL command, probably `drop table`. Or did it work for you on a containerized manticore?

Yes, and this is what I explained also later in my comment. Actually it did work for me locally, but only because I:
a) was mounting the corresponding directory to the host
b) I temporarily switched the path in the ThinkingSphinx env vars to point to the host's directory before running the clear task

What this rake task does actually is just removing the table files: https://github.com/pat/thinking-sphinx/blob/v5.5.1/lib/thinking_sphinx/commands/clear_real_time.rb#L7

We can do this manually. But indeed, maybe we could just issue a _truncate table with reconfigure_ command on a running server? https://manual.manticoresearch.com/Emptying_a_table#Emptying-a-table
I haven't tried this yet.

> I highly suggest we do full reindexing on-prem, i.e. clear the manticore volume. It will not take that much time there and customers will know what's going on. There upgrades are epected to have a little downtime I think.

Well... customers can still know what's going on if we do a partial cleaning. And there will be less downtime. But I actually agree that we might in fact run a full reindexing - because I think it's actually good to run it from time to time, in case the index accidentally gets out of sync.

mayorova commented:
> We can do this manually. But indeed, maybe we could just issue a _truncate table with reconfigure_ command on a running server? https://manual.manticoresearch.com/Emptying_a_table#Emptying-a-table I haven't tried this yet.

So, this works. And it's probably easier. After a new manticore container (with a new config) starts, we can run:

```
TRUNCATE TABLE proxy_rule_core with reconfigure;
```

and then 

```
INDEX_FILTER=proxy_rule_core bundle exec rake ts:rt:index
```

and the change is applied correctly.

jlledom commented:
@mayorova 
> There is an insteresting think I've found: https://edgeapi.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-in_order_of

I think this is the right way to do it. As it's a rails feature we don't have to worry about DB compatibility.

> I think this could be useful if we wanted to order by the list of IDs, as returned by manticore. But, I think it's confusing, because we also support ordering by a selected column etc.

In my opinion it's not confusing, we order by relevance when we receive `nil` as the column to order by, otherwise order by the desired column. This way, when the user does full search it will get the results ordered by relevance unless they click a column.

Right now, the default order column is `position` but the UI doesn't reflect it, so when the user does full search they don't see the position column highlighted, then if they click on the position column, the column is highlighted but the results don't change, because they were already ordered by position before, so IMO it's confusing now.

> And I think this is also what complicates delegating pagination to search engine - the user might request ordering by something else.

Yes, I suggested that but I refuse the idea now.

> What I'm trying to say - let's keep it as it is 😬

I'm fine with that too

jlledom commented:
> > We can do this manually. But indeed, maybe we could just issue a _truncate table with reconfigure_ command on a running server? https://manual.manticoresearch.com/Emptying_a_table#Emptying-a-table I haven't tried this yet.
> 
> So, this works. And it's probably easier. After a new manticore container (with a new config) starts, we can run:
> 
> ```
> TRUNCATE TABLE proxy_rule_core with reconfigure;
> ```

Nice

> 
> ```
> INDEX_FILTER=proxy_rule_core bundle exec rake ts:rt:index
> ```
> 

So I'm lost on this, can we run this command for on premises?

>     1. upgrade manticore (with the new config)
>     2. quickly run the `TRUNCATE TABLE proxy_rule_core with reconfigure;`
>     3. upgrade system

I think we need to also stop Sidekiq, because system is continuously enqueuing reindex jobs every time a model is updated. And those jobs would probably fail if they are performed between steps 1 and 3. So maybe do it like this?

1. Stop sidekiq
2. Deploy system, rollout will update manticore as well, right?
3. Truncate table
4. Reindex
5. Start sidekiq

akostadinov commented:
>> There is an insteresting think I've found: https://edgeapi.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-in_order_of

> I think this is the right way to do it. As it's a rails feature we don't have to worry about DB compatibility.

if this method works, that would be ideal IMO. 

> Deploy system, rollout will update manticore as well, right?

They should be independent I think. But sidekiq will be restarted as soon as system image is changed. 

I think we don't mind if some indexing jobs failed. They should be retried. Also the ones that may fail also are the ones that won't matter because we do full reindexing of that index.

akostadinov commented:
What i we first redeploy manticore and clear the index?

jlledom commented:
> What i we first redeploy manticore and clear the index?

I think it will work as well. Deploying porta only removes the stars from the queries, so it doesn't have huge effects, we can do it at any point of the sequence. We just should ensure the gap between deploying manticore and re index is as small as possible. So I'd say this sequence should work fine:


1. Deploy manticore
2. `TRUNCATE TABLE proxy_rule_core with reconfigure;`
3. `INDEX_FILTER=proxy_rule_core bundle exec rake ts:rt:index`
4. Deploy porta
